### PR TITLE
API-6341-token-validator-v2

### PR DESCRIPTION
**Pull request by jhulbert-va**
# [API-6341](https://vajira.max.gov/browse/API-6341)

### CODE REVIEW

**Comment by jhulbert-va** [comment]
```
$ TEST_PATIENT="32000225" TEST_TOKEN=${token} ./tests/health-apis-token-validator-test.sh

===== AUTHORIZATION VALIDATION =====
SUCCESS -- Authorization Header Does Not Exist is status (401)
SUCCESS -- Authorization Header Is Not Bearer is status (403)

===== STANDARD ACCESS TOKEN VALIDATION =====
SUCCESS -- Token Read R4 Patient is status (200)
SUCCESS -- Token Search R4 Patient By ID is status (200)
SUCCESS -- Token Search R4 Patient By Family Name is status (403)
SUCCESS -- Token Read R4 Unknown Patient is status (403)
SUCCESS -- Token Search R4 Unknown Patient By ID is status (403)
SUCCESS -- Unknown Token Read R4 Patient is status (401)
SUCCESS -- Unknown Token Search R4 Patient By ID is status (401)
SUCCESS -- Scope Type (read/write) cannot be determined is status (500)
SUCCESS -- Token Does Not Have Required Scope is status (403)

===== STATIC ACCESS TOKEN VALIDATION =====
SUCCESS -- Token Read R4 Patient is status (200)
SUCCESS -- Token Search R4 Patient By ID is status (200)
SUCCESS -- Token Search R4 Patient By Family Name is status (403)
SUCCESS -- Token Read R4 Unknown Patient is status (403)
SUCCESS -- Token Search R4 Unknown Patient By ID is status (403)
SUCCESS -- Unknown Token Read R4 Patient is status (401)
SUCCESS -- Unknown Token Search R4 Patient By ID is status (401)
```

File: kong/plugins/health-apis-token-validator/schema.lua [hunk 1]
```diff
@@ -9,6 +9,7 @@ return {
     custom_scope_validation_enabled = {type = "boolean", default = false},
     custom_scope = {type = "string", default = ""},
     api_key = {type = "string"},
-    audience = { type = "string", default = ""}
+    audience = { type = "array", required = true},
+    strict_matching = { type = "string", default = "true"}
```
Comment (bschofield-va) [inline kong/plugins/health-apis-token-validator/schema.lua:13]: it's not clear from the name what "strict matching" applies to. Scopes? Audience? API Key? Is there a better name that might be more clear as to what it actually does?

File: kong/plugins/health-apis-token-validator/handler.lua [hunk 2]
```diff
@@ -39,98 +39,183 @@ function HealthApisTokenValidator:new()
   HealthApisTokenValidator.super.new(self, "health-apis-token-validator")
 end
 
+--
+-- Entrypoint to a kong plugin
+--
 function HealthApisTokenValidator:access(conf)
   HealthApisTokenValidator.super.access(self)
-  kong.log.info("Validating token")
+  kong.log.info("Performing token validation.")
```
Comment (bschofield-va) [inline kong/plugins/health-apis-token-validator/handler.lua:47]: I wonder if we should be more judicious with logging... that line will be printed about 6 million times a day for DQ.

File: kong/plugins/health-apis-token-validator/handler.lua [hunk 3]
```diff
@@ -39,98 +39,183 @@ function HealthApisTokenValidator:new()
...
+  if (requestType == "patient") then
+    return true
+  elseif (requestType == "system" or requestType == "user") then
+    if (launchPatient == nil) then
+      return false
     end
+    return true
   end
+  -- We shouldn't reach here ever, but weirder things have happened.
+  return self:send_response(500, VALIDATE_ERROR)
```
Comment (bschofield-va) [inline kong/plugins/health-apis-token-validator/handler.lua:129]: good failsafe

File: kong/plugins/health-apis-token-validator/handler.lua [hunk 4]
```diff
@@ -39,98 +39,183 @@ function HealthApisTokenValidator:new()
...
-    return client:request_uri(self.conf.verification_url, {
-      method = "GET",
-      ssl_verify = false,
-      headers = headers
-    })
-  end
+
   local body = {
-     aud = self.conf.audience
+     ["aud[]"] = self.conf.audience,
```
Comment (bschofield-va) [inline kong/plugins/health-apis-token-validator/handler.lua:179]: Does this work... I'm under the impression that when multiple audiences are used...
```
aud[]=foo
aud[]=bar
```
vs. 
```
aud[]=foo, bar
```

Did we verify this works?

Comment (jhulbert-va) [reply kong/plugins/health-apis-token-validator/handler.lua:179]: With url encoding, it becomes `aud[]=foo&aud[]=bar&strict=false`. I checked that out in the logs. In addition, I tested this using DQ with multiple audiences.

**Review by bschofield-va — APPROVED** [review]

**Pull request merged**